### PR TITLE
Fix various Uikit memory leaks

### LINKED ISSUE #1463: Autolayout related memory leaks
The heap profiler shows this allocation stack associated with a large number of leaks:

screen shot 2016-11-28 at 8 50 59 am

^^ above shows session with heap snapshots taken on the main menu of the WOCCatalog app (snapshot taken every time a view controller is loaded and unloaded; WOCCatalog needs patch from #1462). There's about a 400KB leak, not entirely autolayout related; leak is worse without fix in #1461.

### LINKED ISSUE #1800: UILabel/UIButton memory leaks in usage of projected WinRT make* APIs
Any time we allocate WinRT-projected objects via make*, we are responsible for freeing them.  However, UILabel and UIButton don't take this into account.

### PR DESCRIPTION
**Pull request by jaredhms**
Fixes #1463.
Fixes #1800.

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/winobjc/1803)

### CODE REVIEW

File: Frameworks/UIKit/UIButton.mm [hunk 1]
```diff
@@ -552,8 +552,7 @@ - (void)setTitleColor:(UIColor*)color forState:(UIControlState)state {
     // ConvertUIColorToWUColor:nil creates a valid WUColor with null comObj
     // which isn't what we want
     if (color) {
-        WUColor* convertedColor = XamlUtilities::ConvertUIColorToWUColor(color);
-        WUXMSolidColorBrush* titleColorBrush = [WUXMSolidColorBrush makeInstanceWithColor:convertedColor];
+        WUXMSolidColorBrush* titleColorBrush = [[WUXMSolidColorBrush makeInstanceWithColor:XamlUtilities::ConvertUIColorToWUColor(color)] autorelease];
         if (titleColorBrush) {
             _states[state].inspectableTitleColor = [titleColorBrush comObj];
```
Comment (DHowett-MSFT) [inline Frameworks/UIKit/UIButton.mm]: 😲 we are literally allocating an objective-c thing that wraps a com thing to get at the com thing and discard the objective-c thing.

what the actual heck

File: samples/WOCCatalog/WOCCatalog/AutoLayoutViewController.m [hunk 2]
```diff
@@ -107,6 +109,44 @@ - (void)viewDidLoad {
 
     [self.view addConstraints: [NSLayoutConstraint constraintsWithVisualFormat:@"|-[label1]-[label2(label1)]-[label3(label1)]-[label4(label1)]-|" options:NO metrics:nil views:NSDictionaryOfVariableBindings(label1, label2, label3, label4)]];
     [self.view addConstraints: [NSLayoutConstraint constraintsWithVisualFormat:@"V:[topLabel]-[label1]-[label2(label1)]-[label3(label1)]-[label4(label1)]-[button]" options:NO metrics:nil views:NSDictionaryOfVariableBindings(topLabel, button, label1, label2, label3, label4)]];
+
+    UIButton* button2 = [UIButton buttonWithType : UIButtonTypeRoundedRect];
+    [button2 setTitle:@"Add a new label" forState:UIControlStateNormal];
```
Comment (DHowett-MSFT) [inline samples/WOCCatalog/WOCCatalog/AutoLayoutViewController.m]: The change in this file does not appear to match anything related to the pull reuqest.

Comment (jaredhms) [inline samples/WOCCatalog/WOCCatalog/AutoLayoutViewController.m]: Autolayout testing

File: Frameworks/UIKit/UIButton.mm (hunk 1 above)
Comment (jaredhms) [inline Frameworks/UIKit/UIButton.mm]: If this becomes a hot path, we can pull in WRL.  

---
In reply to: [97687732](https://github.com/Microsoft/WinObjC/pull/1803#discussion_r97687732) [](ancestors = 97687732)

**Comment by yiyang-msft** [comment]
:shipit:

File: deps/3rdparty/cassowary-0.60/c++/ClSimplexSolver.cc [hunk 3]
```diff
@@ -440,6 +440,9 @@ ClSimplexSolver::RemoveConstraintInternal(const ClConstraint *const pcn)
     cerr << "delete@ " << pexpr << endl;
 #endif
     delete pexpr;
+
+    // Delete the ClAbstractVarable that backs marker, else we leak
+    delete marker.get_pclv();
```
Comment (yiyang-msft) [inline deps/3rdparty/cassowary-0.60/c++/ClSimplexSolver.cc]: >marker [](start = 11, length = 6)

wasn't quite sure how these are used coupled with autolayout, we can sync offline.

File: Frameworks/UIKit/UIButton.mm [hunk 4]
```diff
@@ -552,8 +552,7 @@ - (void)setTitleColor:(UIColor*)color forState:(UIControlState)state {
     // ConvertUIColorToWUColor:nil creates a valid WUColor with null comObj
     // which isn't what we want
     if (color) {
-        WUColor* convertedColor = XamlUtilities::ConvertUIColorToWUColor(color);
-        WUXMSolidColorBrush* titleColorBrush = [WUXMSolidColorBrush makeInstanceWithColor:convertedColor];
+        WUXMSolidColorBrush* titleColorBrush = [[WUXMSolidColorBrush makeInstanceWithColor:XamlUtilities::ConvertUIColorToWUColor(color)] autorelease];
```
Comment (yiyang-msft) [inline Frameworks/UIKit/UIButton.mm]: >autorelease [](start = 138, length = 11)

is there anything preventing us from turning on arc on UIButton/UILabel?

Comment (jaredhms) [inline Frameworks/UIKit/UIButton.mm]: good question; i'll give it a shot

---
In reply to: [97843932](https://github.com/Microsoft/WinObjC/pull/1803#discussion_r97843932) [](ancestors = 97843932)

Comment (yiyang-msft) [inline Frameworks/UIKit/UIButton.mm]: Awesome! Thanks for pushing this!

---
In reply to: [97875917](https://github.com/Microsoft/WinObjC/pull/1803#discussion_r97875917) [](ancestors = 97875917,97843932)

**Comment by yiyang-msft** [comment]
>     [super dealloc];

still need keep this? otherwise, super will not be called? 

---
Refers to: Frameworks/UIKit/UIButton.mm:815 in 5ba191e. [](commit_id = 5ba191e13fc72216926417a12fec543157533cdd, deletion_comment = True)

**Comment by ashvarma** [comment]
:shipit:

**Comment by jaredhms** [comment]
>     [super dealloc];

you can't call dealloc with ARC enabled

---
In reply to: [275424648](https://github.com/Microsoft/WinObjC/pull/1803#issuecomment-275424648) [](ancestors = 275424648)

---
Refers to: Frameworks/UIKit/UIButton.mm:815 in 5ba191e. [](commit_id = 5ba191e13fc72216926417a12fec543157533cdd, deletion_comment = True)

**Pull request merged**